### ISSUE TITLE
Improve popup dictionary styling

### ISSUE THREAD

**Issue by spacehamster** [ripose-jp/Memento#220]
I'd like to see make improvements to Memento's dictionary definition display to bring it more in line with Yomitan. More specifically, I would like to see proper support for Yomichan's compact glossaries, compact tags and better support for structured-content dictionary data. This includes quite a few separate changes:

* In compact glossaries mode, pitch accent entries should be a single line separated by '|'
* In compact glossaries mode, structured content glossaries be on the same line as the definition (part of speech, etc) tags
* In compact glossaries mode, \<ul\> list entries should be on a single line separated by '|'
* In compact tags mode, tags should be omitted if they are included in the previous entry.
* Some dictionaries (such as Jitendex) do not include \<rp\> tags in the ruby markup. Ruby rendering should be improved. Options include stripping \<rt\> tags entirely, highlighting ruby kanji and showing kana on hover as a tooltip, or rendering them inline with brackets.
* Add support for links. Should links open up a new popup or use the existing popup?
* Various styling options are allowed in structured content. Many of them should be ignored due to limitations in QLabel's html rendering engine, but some of them can be implemented.
* Show tooltip when clicking a tag in addition to hover. 
* This is a really minor thing, but I'd like the option to hide the checkboxes.
* Why is there a separate option for "Bullet Points" and "Line Breaks" in the Separate Glossary Entries setting? Structured Content allows \<div\> an \<br\> tags as well as arbitrary Unicode strings for \<li\> list markers. It seems like that choice should left to the dictionary.

This is partially related to anki styling https://github.com/ripose-jp/Memento/issues/213 as many of these points also apply, but because Qt and Anki html rendering have very different capabilities, it seems appropriate they should be handled separately.

Example of yomitan and memento side by side for JMDict and Jitendex

![image](https://github.com/ripose-jp/Memento/assets/3821056/3a2dea56-855d-48dc-acd1-5e068028fcd2)

I can have a go at making a pull request if you're interested.

**Comment by ripose-jp** [comment]
>Some dictionaries (such as Jitendex) do not include `<rp>` tags in the ruby markup. Ruby rendering should be improved. Options include stripping `<rt>` tags entirely, highlighting ruby kanji and showing kana on hover as a tooltip, or rendering them inline with brackets.

All three of these options are desirable, though I like rendering them inline with brackets the least since it gets in the way of readability. Tooltips can be implemented in a hacky way by making tooltip'd text a link and using `linkHovered()` to show a tooltip.

>Add support for links. Should links open up a new popup or use the existing popup?

By links, do you mean clicking a word opens a sub-search of sorts? I'm not against the idea. Opening it in a new popup would be the easiest thing to do since recursive search is already supported. You'd just have to listen for `linkActiviated()` and construct a new popup similarly to know recursive search does so. You can also make it work like clicking kanji in glossaries does, but that's probably more trouble than it's worth. Both are valid approaches.

>Why is there a separate option for "Bullet Points" and "Line Breaks" in the Separate Glossary Entries setting? Structured Content allows `<div>` an `<br>` tags as well as arbitrary Unicode strings for `<li>` list markers. It seems like that choice should left to the dictionary.

Memento is older than Yomitan by several years. Yomitan has changed Yomichan's dictionaries quite a lot, which Memento hasn't kept up with. Very few of Yomichan's dictionaries used structured content, instead opting for plaintext. Most dictionaries used `\n` to separate different definitions under the same glossary. Older versions of Memento would turn these into bulleted lists. Calvin didn't like the unused white space to the left of bullets, so he made #64.

I believe the choice is mostly left to the dictionary if it's using structured content. These settings have an outsized effect on plaintext dictionaries.

---

If I didn't comment on something, it's because I'm okay with the change/improvement. If you want to work on some of them and make PR's, by all means go ahead.

**Comment by spacehamster** [comment]
> By links, do you mean clicking a word opens a sub-search of sorts?

Structured content allows dictionaries to specify \<a\> tags which can be either sub-search or links to external websites. Yomitan uses the same popup, but I have no issues with the way Memento currently opens a new popup for recursive search.

The structured content format looks like this
```
{
  "tag": "a",
  "lang": "ja",
  "href": "?query=日曜&wildcards=off",
  "content": [ <snip> ]
}
```

![image](https://github.com/ripose-jp/Memento/assets/3821056/7dce99ad-08ce-4470-9dc0-a90bff521d41)

**Comment by ripose-jp** [comment]
Makes sense to me. I'd say reuse the code for recursive search since that's going to be really easy.

**Comment by spacehamster** [comment]
I'm not able to do anymore work on this for the foreseeable future. I'll post what I've done so far in case it's useful. 

I've added
* compact pitch accents: https://github.com/spacehamster/Memento/tree/pitch_accent
* compact tags: https://github.com/spacehamster/Memento/tree/compact_tags
* compact glosseries https://github.com/spacehamster/Memento/tree/compact_glossary

![image](https://github.com/ripose-jp/Memento/assets/3821056/af98d2d1-b876-4bde-8921-ea679f3f2f3d)

The approach uses flow layout for most of the changes. There are some layout issues with this method that I don't think can be completely eliminated, so I'm not really satisfied with it.

![image](https://github.com/ripose-jp/Memento/assets/3821056/941cf321-a192-48a3-80d4-70f4c742003e)

The fundamental problem is that the html subset used by QLabels is extremely limited and makes it difficult to do proper layout. Had time allowed, I had planned to evaluate the feasibility of litehtml (html+css engine with no renderer included) to do all the layout work and render either each term widget as a html document or the entire definition widget as a html document. The main potential issues I foresaw was performance and licensing. Litehtml is BSD-3 and there are qt bindings [here](https://codereview.qt-project.org/admin/repos/playground/qlitehtml,general) licensed under LicenseRef-Qt-Commercial OR GPL-3.0-only WITH Qt-GPL-exception-1.0.

**Comment by ripose-jp** [comment]
Thanks for the work you've done. I'll look into integrating what you have.

**Comment by ripose-jp** [comment]
I think I've addressed most of the concerns expect for compact mode.

* `<rub>` and `<rp>` tags are rendered as tool tips when hovering over the characters.
* Links work for recursive searches and opening external web sites.
* Styling isn't perfect, but much better. See below.
* Tool tips are shown immediately when clicking tags.
* Add to Anki checkboxes can optionally be hidden

<img width="400" height="893" alt="Image" src="https://github.com/user-attachments/assets/fb65e601-198b-4844-a0bc-82b72b5fa6e7" />

<img width="399" height="808" alt="Image" src="https://github.com/user-attachments/assets/fb299dde-0b6f-41ff-b583-4d2860803a3d" />

**Comment by ripose-jp** [comment]
The latest changes make Jitendex almost perfect. It does seem to come at a performance cost, but it seems to be on the render side rather than generating markup itself. It's not so slow that I think it would bother anyone though.
<img width="417" height="1297" alt="Image" src="https://github.com/user-attachments/assets/c0e086e3-2685-478b-b94c-0e942cd43cb1" />

**Comment by TakeshiTogo** [comment]
It’s beautiful. 🥹
